### PR TITLE
Make eval_place iterate instead of recurse

### PR DESCRIPTION
**Pull request by spastorino**
r? @oli-obk

### CODE REVIEW

File: src/librustc_mir/interpret/place.rs [hunk 1]
```diff
@@ -611,35 +611,37 @@ where
...
+            for proj in place_projection {
+                place = self.place_projection(place, &proj.elem)?
             }
 
-            _ => self.eval_place_to_mplace(mir_place)?.into(),
-        };
+            Ok(place)
+        })?;
 
         self.dump_place(place.place);
```
Comment (oli-obk) [inline src/librustc_mir/interpret/place.rs]: move this into the `iterate` closure

File: src/librustc_mir/interpret/place.rs [hunk 2]
```diff
@@ -611,35 +611,37 @@ where
...
         Ok(place)
```
Comment (oli-obk) [inline src/librustc_mir/interpret/place.rs]: remove this

File: src/librustc_mir/interpret/place.rs [hunk 3]
```diff
@@ -611,35 +611,37 @@ where
...
-                let place = self.eval_place(&proj.base)?;
-                self.place_projection(place, &proj.elem)?
+            for proj in place_projection {
+                place = self.place_projection(place, &proj.elem)?
             }
 
-            _ => self.eval_place_to_mplace(mir_place)?.into(),
-        };
+            Ok(place)
+        })?;
```
Comment (oli-obk) [inline src/librustc_mir/interpret/place.rs]:
```suggestion
        })
```

**Review by oli-obk — CHANGES_REQUESTED** [review]

File: src/librustc_mir/interpret/place.rs [hunk 4]
```diff
@@ -611,35 +611,37 @@ where
...
-        let place = match *mir_place {
-            Base(PlaceBase::Local(mir::RETURN_PLACE)) => match self.frame().return_place {
-                Some(return_place) =>
-                    // We use our layout to verify our assumption; caller will validate
-                    // their layout on return.
-                    PlaceTy {
-                        place: *return_place,
-                        layout: self.layout_of(self.monomorphize(self.frame().mir.return_ty())?)?,
+
+        let place = mir_place.iterate(|place_base, place_projection| {
```
Comment (oli-obk) [inline src/librustc_mir/interpret/place.rs]:
```suggestion
        mir_place.iterate(|place_base, place_projection| {
```

**Comment by oli-obk** [comment]
@spastorino you have a tidy failure.

you can run tidy with `./x.py test src/tools/tidy`

**Comment by spastorino** [comment]
I think that shouldn't require any special rights, I'm trying to r- a failing PR that I've created /cc @rust-lang/infra. Maybe we can improve this?.

File: src/librustc_mir/interpret/place.rs [hunk 5]
```diff
@@ -609,40 +609,40 @@ where
...
+                    None => return err!(InvalidNullPointerUsage),
                 },
-                layout: self.layout_of_local(self.frame(), local, None)?,
-            },
+                PlaceBase::Local(local) => PlaceTy {
+                    // This works even for dead/uninitialized locals; we check further when writing
+                    place: Place::Local { frame: self.cur_frame(), local: *local },
+                    layout: self.layout_of_local(self.frame(), *local, None)?,
+                },
+                _ => self.eval_place_to_mplace(mir_place)?.into(),
```
Comment (oli-obk) [inline src/librustc_mir/interpret/place.rs]: This is the same issue as the other PR I guess?

Comment (spastorino) [reply src/librustc_mir/interpret/place.rs]: yes, waiting for https://github.com/rust-lang/rust/pull/61104 to be in

**Comment by spastorino** [comment]
When https://github.com/rust-lang/rust/pull/61104 is merged this is going to be ready to be merged.

File: src/librustc_mir/interpret/place.rs [hunk 6]
```diff
@@ -609,40 +609,40 @@ where
...
-                        layout: self.layout_of(self.monomorphize(self.frame().mir.return_ty())?)?,
-                    },
-                None => return err!(InvalidNullPointerUsage),
-            },
-            Base(PlaceBase::Local(local)) => PlaceTy {
-                // This works even for dead/uninitialized locals; we check further when writing
-                place: Place::Local {
-                    frame: self.cur_frame(),
-                    local,
+                    {
```
Comment (RalfJung) [inline src/librustc_mir/interpret/place.rs]: I think we usually put the `{` on the same line as the `=>`.

**Pull request merged**